### PR TITLE
community:Updating Konko Integration

### PR DESCRIPTION
**Pull request by shivanimodi16**
# Description

This PR introduces update to Konko Integration with LlamaIndex.

1. Konko's Completion Endpoint Addition: Integration of completion endpoint to utilize a new set of models hosted on Konko.
2. Updated Documentation: Added jupyter notebook to reflect these new changes, providing clear guidance on utilizing the new features and ensuring seamless integration.

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new notebook (that tests end-to-end)

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/llms/konko.py [hunk 1]
```diff
@@ -534,7 +536,8 @@ async def _acomplete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
             stream=False,
             **all_kwargs,
         )
-        text = response["choices"][0]["text"]
+        print(response)
```
Comment (dalazx) [inline llama_index/llms/konko.py]: are we sure we want to leave this `print` here?

File: llama_index/llms/konko.py [hunk 2]
```diff
@@ -130,12 +126,50 @@ def _get_model_name(self) -> str:
...
+        return context_size
+
     @property
     def metadata(self) -> LLMMetadata:
         return LLMMetadata(
-            context_window=konko_modelname_to_contextsize(self._get_model_name()),
+            context_window=self._konko_modelname_to_contextsize(self._get_model_name()),
             num_output=self.max_tokens,
-            is_chat_model=True,
+            is_chat_model=is_chat_model(self._get_model_name()),
```
Comment (dalazx) [inline llama_index/llms/konko.py]: I think we may be doing some double work here: he request /v1/models once in `_konko_modelname_to_contextsize`, and then again in `is_chat_model`. we could easily turn this into a single request. we could either expand the dict we return in `_create_model_context_length_dict` or even better introduce a dataclass that would hold all relevant attributes of a model, and many such instances would be stored in the resulting dict.

File: llama_index/llms/konko.py [hunk 3]
```diff
@@ -222,52 +257,28 @@ def _stream_chat(
...
                 max_retries=self.max_retries,
                 messages=message_dicts,
                 stream=True,
                 **all_kwargs,
             ):
-                if len(response["choices"]) == 0 and response.get("prompt_annotations"):
-                    # When asking a stream response from the Azure OpenAI API
-                    # you first get an empty message with the content filtering
-                    # results. Ignore this message
+                if len(response.choices) == 0 and response.prompt_annotations:
```
Comment (dalazx) [inline llama_index/llms/konko.py]: curious whether we need to check the openai version here as well.

File: llama_index/llms/konko.py [hunk 4]
```diff
@@ -355,8 +370,8 @@ def gen() -> CompletionResponseGen:
                 stream=True,
                 **all_kwargs,
             ):
-                if len(response["choices"]) > 0:
-                    delta = response["choices"][0]["text"]
+                if len(response.choices) > 0:
```
Comment (dalazx) [inline llama_index/llms/konko.py]: and here.

File: llama_index/llms/konko.py [hunk 5]
```diff
@@ -481,34 +499,18 @@ async def gen() -> ChatResponseAsyncGen:
                 stream=True,
                 **all_kwargs,
             ):
-                if len(response["choices"]) > 0:
-                    delta = response["choices"][0]["delta"]
+                if len(response.choices) > 0:
```
Comment (dalazx) [inline llama_index/llms/konko.py]: and here.

File: llama_index/llms/konko.py [hunk 6]
```diff
@@ -534,7 +536,8 @@ async def _acomplete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
             stream=False,
             **all_kwargs,
         )
-        text = response["choices"][0]["text"]
+        print(response)
+        text = response.choices[0].text
```
Comment (dalazx) [inline llama_index/llms/konko.py]: and here.

File: llama_index/llms/konko_utils.py [hunk 7]
```diff
@@ -145,15 +158,19 @@ def to_openai_message_dicts(messages: Sequence[ChatMessage]) -> List[dict]:
...
-    additional_kwargs.pop("content", None)
-
+    if is_openai_v1():
+        role = message_dict.role
+        content = message_dict.content
+    else:
+        role = message_dict["role"]
+        content = message_dict.get("content", None)
+    additional_kwargs = {
+        attr: getattr(message_dict, attr)
```
Comment (dalazx) [inline llama_index/llms/konko_utils.py]: how does this work with a `dict` returned by the version 0?

File: llama_index/llms/konko_utils.py [hunk 8]
```diff
@@ -219,12 +236,27 @@ async def acompletion_with_retry(
...
 ) -> Any:
     """Use tenacity to retry the async completion call."""
+    try:
+        import konko
+
+    except ImportError:
+        raise ValueError(
+            "Could not import konko python package. "
+            "Please install it with `pip install konko`."
+        )
```
Comment (dalazx) [inline llama_index/llms/konko_utils.py]: this is quite repetitive. we should extract a function and reuse.

File: llama_index/llms/konko.py [hunk 9]
```diff
@@ -562,8 +565,8 @@ async def gen() -> CompletionResponseAsyncGen:
                 stream=True,
                 **all_kwargs,
             ):
-                if len(response["choices"]) > 0:
-                    delta = response["choices"][0]["text"]
+                if len(response.choices) > 0:
```
Comment (dalazx) [inline llama_index/llms/konko.py]: same here.

File: llama_index/llms/konko.py (hunk 2 above)
Comment (shivanimodi16) [reply llama_index/llms/konko.py]: Makes sense! Did the changes

Comment (dalazx) [reply llama_index/llms/konko.py]: hm, not sure I see them.

Comment (shivanimodi16) [reply llama_index/llms/konko.py]: Just pushed the code.

File: llama_index/llms/konko.py (hunk 3 above)
Comment (shivanimodi16) [reply llama_index/llms/konko.py]: Done

File: llama_index/llms/konko.py (hunk 4 above)
Comment (shivanimodi16) [reply llama_index/llms/konko.py]: done

File: llama_index/llms/konko.py (hunk 5 above)
Comment (shivanimodi16) [reply llama_index/llms/konko.py]: done

File: llama_index/llms/konko.py (hunk 6 above)
Comment (shivanimodi16) [reply llama_index/llms/konko.py]: done

File: llama_index/llms/konko_utils.py (hunk 7 above)
Comment (shivanimodi16) [reply llama_index/llms/konko_utils.py]: done

File: llama_index/llms/konko_utils.py (hunk 8 above)
Comment (shivanimodi16) [reply llama_index/llms/konko_utils.py]: done

File: llama_index/llms/konko.py (hunk 9 above)
Comment (shivanimodi16) [reply llama_index/llms/konko.py]: done

File: llama_index/llms/konko.py [hunk 10]
```diff
@@ -130,18 +137,70 @@ def _get_model_name(self) -> str:
     def class_name(cls) -> str:
         return "Konko_LLM"
 
+    def _create_model_info_dict(self) -> dict:
```
Comment (dalazx) [inline llama_index/llms/konko.py]:
```suggestion
    def _create_model_info_dict(self) -> Dict[str, ModelInfo]:
```

File: llama_index/llms/konko.py [hunk 11]
```diff
@@ -130,18 +137,70 @@ def _get_model_name(self) -> str:
...
+        return model_info.is_chat_model
+
     @property
     def metadata(self) -> LLMMetadata:
+        model_info = self.model_info_dict.get(self._get_model_name())
+        if model_info is None:
+            raise ValueError(
+                f"Unknown model: {self._get_model_name()}. Please provide a valid Konko model name."
+                "Known models are: " + ", ".join(self.model_info_dict.keys())
+            )
```
Comment (dalazx) [inline llama_index/llms/konko.py]: Minor. This piece also seems repetitive. We could extract a method and reuse.

**Review by dalazx — APPROVED** [review]
one minor comment. great job!

**Pull request merged**